Date: 2005-09-20
Initial Package Version: 2.9.7
Upstream Status: Not yet submitted
Origin: Miklos Vajna <vmiklos@frugalware.org>
Description: Replaces pacman -Ss foo['s segfault with an error message.
diff -Naur pacman-2.9.7.orig/src/db.c pacman-2.9.7/src/db.c
--- pacman-2.9.7.orig/src/db.c	2005-09-20 00:33:28.000000000 +0200
+++ pacman-2.9.7/src/db.c	2005-09-20 00:45:36.000000000 +0200
@@ -660,8 +660,6 @@
 	}
 
 	for(i = needles; i; i = i->next) {
-		char *targ = strdup(i->data);
-		strtoupper(targ);
 		for(j = cache; j; j = j->next) {
 			pkginfo_t *pkg = (pkginfo_t*)j->data;
 			char *haystack;
@@ -669,7 +667,7 @@
 			/* check name */
 			haystack = strdup(pkg->name);
 			strtoupper(haystack);
-			if(reg_match(haystack, targ)) {
+			if(reg_match(haystack, i->data)) {
 				match = 1;
 			}
 			FREE(haystack);
@@ -678,7 +676,7 @@
 			if(!match) {
 				haystack = strdup(pkg->desc);
 				strtoupper(haystack);
-				if(reg_match(haystack, targ)) {
+				if(reg_match(haystack, i->data)) {
 					match = 1;
 				}
 				FREE(haystack);
@@ -692,7 +690,7 @@
 					for(m = info->provides; m; m = m->next) {
 						haystack = strdup(m->data);
 						strtoupper(haystack);
-						if(reg_match(haystack, targ)) {
+						if(reg_match(haystack, i->data)) {
 							match = 1;
 						}
 						FREE(haystack);
@@ -707,7 +705,7 @@
 				printf("\n");
 			}
 		}
-		FREE(targ);
+		FREE(i->data);
 	}
 }
 
diff -Naur pacman-2.9.7.orig/src/util.c pacman-2.9.7/src/util.c
--- pacman-2.9.7.orig/src/util.c	2005-09-20 00:33:28.000000000 +0200
+++ pacman-2.9.7/src/util.c	2005-09-20 00:51:52.000000000 +0200
@@ -403,7 +403,11 @@
 	int result;
 	regex_t reg;
 
-	regcomp(&reg, pattern, REG_EXTENDED | REG_NOSUB);
+	if (regcomp(&reg, pattern, REG_EXTENDED | REG_NOSUB) != 0)
+	{
+		fprintf(stderr, "error: %s is not a valid regular expression.\n", pattern);
+		exit(1);
+	}
 	result = regexec(&reg, string, 0, 0, 0);
 	regfree(&reg);
 	return(!(result));
